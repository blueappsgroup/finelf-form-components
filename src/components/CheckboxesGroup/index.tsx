/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { ReactElement, FC, ChangeEvent } from 'react'
import { FieldArray, useFormikContext } from 'formik'

import CheckboxField from '../FormInput/CheckboxField'

type CheckboxesGroupTypes = {
  name: string
  children: any
}

const CheckboxesGroup: FC<CheckboxesGroupTypes> = ({ name, children }) => {
  const { setValues } = useFormikContext()
  const hasManyCheckboxes = Array.isArray(children)
  const checkboxesNames: string[] = hasManyCheckboxes
    ? children && children.map((item: any) => item.props.name)
    : [children.props.name]

  const handleOnChange = (e: ChangeEvent<HTMLInputElement>): void => {
    const fieldsToUpdate = checkboxesNames.reduce((acc, item) => {
      acc[item] = e.target.value === 'true' ? false : true
      return acc
    }, {} as { [key: string]: boolean | undefined })

    setValues({ [name]: fieldsToUpdate })
  }

  return (
    <FieldArray
      name={name}
      render={(): ReactElement => (
        <>
          <CheckboxField
            id={`${name}.selectAll`}
            name={`${name}.selectAll`}
            label="Select all"
            onChange={handleOnChange}
          />
          {children && !hasManyCheckboxes && (
            <CheckboxField
              {...children.props}
              name={`${name}.${children.props.name}`}
            />
          )}
          {children &&
            hasManyCheckboxes &&
            children.map((item: any) => (
              <CheckboxField
                {...item.props}
                key={item.props.name}
                name={`${name}.${item.props.name}`}
              />
            ))}
        </>
      )}
    />
  )
}

export default CheckboxesGroup
