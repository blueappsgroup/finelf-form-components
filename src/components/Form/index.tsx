import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
