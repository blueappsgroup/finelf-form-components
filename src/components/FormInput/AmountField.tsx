import React, { FC } from 'react'
import { Field } from 'formik'

import { RangeFieldProps } from '../../types'
import { BaseRangeField } from './base'

const AmountField: FC<RangeFieldProps> = ({
  id,
  label = 'Kwota pożyczki',
  name,
  required = false,
  showError = true,
  showMinMax = true,
  min = 300,
  max = 10000,
  step = 100,
  value = 2000,
  suffix = 'PLN',
  unit = 'zł',
}) => (
  <Field
    required={required}
    showError={showError}
    showMinMax={showMinMax}
    id={id}
    name={name}
    component={BaseRangeField}
    label={label}
    type="text"
    min={min}
    max={max}
    step={step}
    value={value}
    suffix={suffix}
    unit={unit}
  />
)

AmountField.defaultProps = {
  name: 'amount',
}

export default AmountField
