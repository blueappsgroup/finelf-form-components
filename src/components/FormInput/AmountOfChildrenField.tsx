import React, { ReactElement } from 'react'
import { Field } from 'formik'

import BaseField from './base'
import { validateAmountOfChildren } from './validate'
import { AmountOfChildrenFieldProps } from '../../types'

const AmountOfChildrenField: (
  props: AmountOfChildrenFieldProps
) => ReactElement = ({
  id,
  label,
  placeholder = 'AmountOfChildren',
  errorMsg = 'Podana ilość jest nieprawidłowa',
  name = 'amount_of_children',
  required = false,
  showError,
}) => (
  <Field
    required={required}
    showError={showError}
    id={id}
    name={name}
    component={BaseField}
    label={label}
    placeholder={placeholder}
    type="text"
    validate={validateAmountOfChildren(errorMsg, required)}
  />
)

export default AmountOfChildrenField
