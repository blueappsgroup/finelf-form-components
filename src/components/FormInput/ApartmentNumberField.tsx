import React from 'react'
import TextField from './TextField'
import { validateHouseNumber } from './validate'

const ApartmentNumberField = () => (
  <TextField
    required
    label="Numer domu"
    placeholder="Numer domu"
    showError
    name="house_number"
    validate={validateHouseNumber}
  />
)

export default ApartmentNumberField
