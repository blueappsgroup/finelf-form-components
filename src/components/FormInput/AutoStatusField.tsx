import { Field } from 'formik'
import React, { ReactElement } from 'react'
import { SelectFieldProps } from '../../types'

import { BaseSelectField } from './base'

const options = {
  yes: 'tak',
  no: 'nie',
}

const AutoStatusField: (props: SelectFieldProps) => ReactElement = ({
  id,
  label,
  name = 'auto_status',
  required = false,
  showError,
}) => (
  <Field
    required={required}
    showError={showError}
    id={id}
    name={name}
    component={BaseSelectField}
    label={label}
    type="select"
    options={options}
  />
)

export default AutoStatusField
