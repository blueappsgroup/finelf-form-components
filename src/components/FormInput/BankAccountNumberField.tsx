import React, { FC } from 'react'
import { Field } from 'formik'

import { TextFieldProps } from '../../types'
import BaseField from './base'
import { validateBankAccountNumber } from './validate'

const BankAccountNumberField: FC<TextFieldProps> = ({
  id,
  label = 'Numer konta bankowego',
  placeholder,
  name,
  required = false,
  validate = validateBankAccountNumber,
  showError = true,
}) => (
  <Field
    id={id}
    label={label}
    placeholder={placeholder}
    component={BaseField}
    required={required}
    showError={showError}
    name={name}
    validate={validate(required)}
  />
)

BankAccountNumberField.defaultProps = {
  name: 'bank_account_number',
}

export default BankAccountNumberField
