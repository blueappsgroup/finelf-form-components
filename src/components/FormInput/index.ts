export { default as EmailField } from './EmailField'
export { default as TextField } from './TextField'
export { default as TelephoneField } from './TelephoneField'
export { default as CheckboxField } from './CheckboxField'
export { default as FirstNameField } from './FirstNameField'
export { default as LastNameField } from './LastNameField'
export { default as PeselField } from './PeselField'
export { default as HouseNumberField } from './HouseNumberField'
export { default as ApartmentNumberField } from './ApartmentNumberField'
export { default as CityNameField } from './CityNameField'
export { default as StreetNameField } from './StreetNameField'
export { default as ZipCodeField } from './ZipCodeField'
export { default as CompanyNameField } from './CompanyNameField'
export { default as CompanyTaxField } from './CompanyTaxField'
export { default as BankAccountNumberField } from './BankAccountNumberField'
export { default as IDCardNumberField } from './IDCardNumberField'
export { Row } from './base'
