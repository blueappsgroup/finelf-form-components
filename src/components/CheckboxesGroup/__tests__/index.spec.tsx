import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import { CheckboxField } from '../../FormInput'
import CheckboxesGroup from '../index'

describe('base <CheckboxesGroup />', () => {
  const onSubmit = jest.fn()
  const setupWrapper = ({
    formId = 'testForm',
    checkboxFieldLabel = 'test checkbox',
    checkboxesGroupName = 'checkboxes_group',
    checkboxesGroupFirstCheckboxName = 'checkbox1',
    checkboxesGroupSecondCheckboxName = 'checkbox2',
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  }): any => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <CheckboxesGroup name={checkboxesGroupName}>
          <CheckboxField
            name={checkboxesGroupFirstCheckboxName}
            label={checkboxFieldLabel}
          />
          <CheckboxField
            name={checkboxesGroupSecondCheckboxName}
            label={checkboxFieldLabel}
          />
        </CheckboxesGroup>
      </Form>
    )
    const { container } = wrapper

    const checkboxesGroup = container.querySelector(
      `[name="${checkboxesGroupName}.selectAll"]`
    )

    return {
      checkboxesGroup,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { checkboxesGroup } = setupWrapper({})

    expect(checkboxesGroup).toBeTruthy()
    expect(checkboxesGroup).toMatchSnapshot()
  })

  it('CheckboxesGroup field handleOnChange save values to sessionStorage', async () => {
    const { checkboxesGroup } = setupWrapper({})

    await act(async () => {
      fireEvent.click(checkboxesGroup)
    })

    await act(async () => {
      fireEvent.mouseOut(checkboxesGroup)
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm'))
        .checkboxes_group.selectAll
    ).toBe(true)
    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm'))
        .checkboxes_group.checkbox1
    ).toBe(true)
    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm'))
        .checkboxes_group.checkbox2
    ).toBe(true)
    expect(checkboxesGroup.checked).toBe(true)
  })
})
