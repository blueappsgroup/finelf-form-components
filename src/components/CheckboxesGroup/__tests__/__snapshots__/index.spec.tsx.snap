// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base <CheckboxesGroup /> matches snapshot 1`] = `
<input
  class="sc-iktFzd bNeliV"
  id="checkboxes_group.selectAll"
  label="Zaznacz wszystkie zgody"
  name="checkboxes_group.selectAll"
  type="checkbox"
  value="false"
/>
`;
